_ap/17/index.adoc

```diff
@@ -0,0 +1,62 @@
+---
+num: 17
+tags:
+  - "tls"
+  - "certificates"
+  - "dns"
+  - "Letsencrypt"
+title: "Automatic management of certificates per data plane cluster using Letsencrypt"
+status: "Accepted"
+authors:
+  - "Manyanda Chitimbo"
+---
+
+## Intent
+
+Define a pattern for the automatic management of certificates per data plane cluster using Letsencrypt
+
+## Motivation
+
+When a managed service is deployed on the data plane cluster its routes have be reachable via TLS.
+
+It means that a managed service a way to the clients to verify and validate the TLS connection with the related handshake with the service exposed outside of the data plane cluster.
+
+## Applicability
+
+This pattern applies to any service that wants a managed service routes to be reachable via TLS
+
+## Description
+
+A managed service will use https://Letsencrypt.org[Letsencrypt] to managed certificates on the fly i.e automatic and this removing the need for manual intervention and the toil related to this.
+
+To avoid sharing of certificates between data planes, managed service implementors are recommended to generate a wildcard certificate per data plane.
+
+As an example, if the root domain is `myservice.bf2.dev`, then a wildcard certificate will be generated for each data plane cluster under a subdomain `<cluster-id>.myservice.bf2.dev`.
```

Comment:
I forgot whether and what we said about exposing the cluster id in the URL. It does contradict SDR 1. I'll add it tot he conversastion in the arch call.

Reply:
BTW it is unclear what cluster id refers to. Make it more explicit (e.g. Kubernetes cluster id). I thought it was the kafka cluster id for a few minutes until I chcked ADR-89

Reply:
> I forgot whether and what we said about exposing the cluster id in the URL. It does contradict SDR 1. I'll add it tot he conversastion in the arch call.

> BTW it is unclear what cluster id refers to. Make it more explicit (e.g. Kubernetes cluster id). I thought it was the kafka cluster id for a few minutes until I chcked ADR-89

Thanks for raising it. Non, it is not the Kafka id but the data plane cluster id as we need a wildcard cert per data plane cluster.

I'll make it clear.

Reply:
> I forgot whether and what we said about exposing the cluster id in the URL. It does contradict SDR 1. I'll add it tot he conversastion in the arch call.

Thanks, we are using the cluster id for now in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1504 but we could as well envision to generate a unique slug per cluster and use it in the URL. 
@emmanuelbernard do you think we'll get a reply from the call soonish enough for us to go ahead with the change in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1504? Thanks

/cc @pb82

Reply:
@emmanuelbernard permit me to bring a bit of precision on 

> Thanks, we are using the cluster id for now in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1504 but we could as well envision to generate a unique slug per cluster and use it in the URL.

We need a unique identifier for the shard and the `OCM` cluster id was choosen since its unique. 
For example for the `kafka.bf2.dev` domain, and the cluster id `123`, then the base domain for this cluster will be `123.kafka.bf2.dev` and we'll generate a wildcard certificate for it `*.123.kafka.bf2.dev`

When kafkas are deployed on this data plane cluster the kafka urls will be of the form `xxxxx.123.kafka.bf2.dev` e.g the admin server url can be `admin-server-kafka-cf-ho-a---r-kce--otg` 

Going back to what I said above

> We need a unique identifier for the shard and the `OCM` cluster id was choosen since its unique. 

If we deem that the cluster-id shouldn't be used as the shard1, we could envision to generate a unique string that can be used as a shard id. We could use the logical cluster id (the PK in KFM database for the given cluster) or we could generate a totally different shard id that's unique and totally opaque without any meaning. 

Let me know what you think and I'll update the AP and reflects the changes in the ADR-89.

Reply:
The opaque proposal does not solve the core reason that the URL is no longer location agnostic at least if we make the assumption that the certificate remains only on a given data plane cluster and does not travel.

Reply:
Thanks. I think I see the issue now e.g when a kafka is migrated from one OSD to another...